src/stages/tx_lookup.rs

```
@@ -98,17 +98,53 @@ where
         })
     }
 
-    async fn unwind<'tx>(
-        &self,
-        tx: &'tx mut RwTx,
-        input: crate::stagedsync::stage::UnwindInput,
-    ) -> anyhow::Result<()>
+    async fn unwind<'tx>(&self, tx: &'tx mut RwTx, input: UnwindInput) -> anyhow::Result<()>
     where
         'db: 'tx,
     {
-        let _ = tx;
-        let _ = input;
-        todo!()
+        if input.unwind_to >= input.stage_progress {
+            return Ok(());
+        }
+
+        let mut bodies_cursor = tx.mutable_cursor(&tables::BlockBody).await?;
+        let mut tx_hash_cursor = tx.mutable_cursor(&tables::TxLookup).await?;
+        let mut block_txs_cursor = tx.cursor(&tables::BlockTransaction).await?;
+
+        let mut start_block_number = [0; 8];
+
+        U64::from(input.unwind_to + 1).to_big_endian(&mut start_block_number);
+
+        info!(
+            "Started Tx Lookup Unwind, from: {} to: {}",
+            input.stage_progress, input.unwind_to
+        );
+
+        let walker_block_body = txdb::walk(&mut bodies_cursor, &start_block_number, 0);
+        pin!(walker_block_body);
+
+        while let Some((_, ref block_body_value)) = walker_block_body.try_next().await? {
+            let body_rpl = rlp::decode::<BodyForStorage>(block_body_value)?;
+            let (tx_count, tx_base_id) = (body_rpl.tx_amount, body_rpl.base_tx_id);
+            let tx_base_id_as_bytes = tx_base_id.to_be_bytes();
+
+            let walker_block_txs = txdb::walk(&mut block_txs_cursor, &tx_base_id_as_bytes, 0);
+            pin!(walker_block_txs);
+
+            let mut num_txs = 1;
+
+            while let Some((_tx_key, ref tx_value)) = walker_block_txs.try_next().await? {
+                if num_txs > tx_count {
+                    break;
+                }
+
+                let hashed_tx_data = hash_data(tx_value);
+                if let Ok(Some((key, value))) = tx_hash_cursor.seek(hashed_tx_data.as_ref()).await {
```

Why do you suppress the error here?

---

good catch